### PR TITLE
feat(core): interaction hooks

### PR DESCRIPTION
**Pull request by gao-sun**
## Summary

implement `koaInteractionHooks()` middleware function to trigger hooks after interaction when needed

## Testing

unit tests, pending integration tests after new interaction API is adopted

### CODE REVIEW

**Review by simeng-li — CHANGES_REQUESTED** [review]
hold for PR #2686. will rebase after the PR is merged.

**Review by simeng-li — APPROVED** [review]

File: packages/core/src/routes/interaction/index.ts [hunk 1]
```diff
@@ -70,18 +65,19 @@ export default function interactionRoutes<T extends AnonymousRouter>(
         profile: profileGuard.optional(),
       }),
     }),
+    koaInteractionSIE(),
```
Comment (gao-sun) [inline packages/core/src/routes/interaction/index.ts:68]:
```suggestion
    koaInteractionSie(),
```

File: packages/core/src/routes/interaction/middleware/koa-interaction-sie.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  WithInteractionSIEContext<ContextT>,
+  ResponseT
+> {
+  return async (ctx, next) => {
+    const { interactionDetails } = ctx;
+
+    const signInExperience = await getSignInExperienceForApplication(
+      typeof interactionDetails.params.client_id === 'string'
+        ? interactionDetails.params.client_id
+        : undefined
```
Comment (gao-sun) [inline packages/core/src/routes/interaction/middleware/koa-interaction-sie.ts:23]: conditional

File: packages/core/src/routes/interaction/verifications/mandatory-user-profile-validation.ts [hunk 3]
```diff
@@ -71,11 +70,10 @@ const getMissingProfileBySignUpIdentifiers = ({
 };
 
 export default async function validateMandatoryUserProfile(
-  ctx: Context,
-  provider: Provider,
+  ctx: WithInteractionSIEContext<Context>,
```
Comment (gao-sun) [inline packages/core/src/routes/interaction/verifications/mandatory-user-profile-validation.ts:73]:
```suggestion
  ctx: WithInteractionSieContext<Context>,
```

File: packages/schemas/src/models/hooks.ts [hunk 4]
```diff
@@ -2,12 +2,29 @@ import { generateStandardId } from '@logto/core-kit';
...
+  event: HookEvent;
+  createdAt: string;
+  sessionId?: string;
+  userAgent?: string;
+  userId?: string;
+  user?: Pick<User, typeof userInfoSelectFields[number]>;
+  application?: Pick<Application, 'id' | 'type' | 'name' | 'description'>;
+  connectors?: Array<
+    Pick<Connector, 'id'> & Pick<Connector['metadata'], 'name'> & Record<string, unknown>
+  >;
```
Comment (gao-sun) [inline packages/schemas/src/models/hooks.ts:25]:
```suggestion
```

File: packages/core/src/routes/interaction/index.ts (hunk 1 above)
Comment (simeng-li) [reply packages/core/src/routes/interaction/index.ts]: updated

File: packages/core/src/routes/interaction/verifications/mandatory-user-profile-validation.ts (hunk 3 above)
Comment (simeng-li) [reply packages/core/src/routes/interaction/verifications/mandatory-user-profile-validation.ts]: updated

**Review by simeng-li — APPROVED** [review]
LGTM

File: packages/core/src/libraries/hook.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+  if (!userId) {
+    return;
+  }
+
+  const interactionPayload = getInteractionStorage(details.result);
+  const { event } = interactionPayload;
+
+  const hookEvent = eventToHook[event];
+  const { rows } = await modelRouters.hook.client.readAll();
```
Comment (wangsijie) [inline packages/core/src/libraries/hook.ts:37]: Why not add a "where" condition here?

**Review by wangsijie — APPROVED** [review]
Overall LGTM

File: packages/core/src/index.ts [hunk 6]
```diff
@@ -1,23 +1,16 @@
...
-    await configDotEnv();
-    await envSet.load();
-    const app = new Koa({
-      proxy: envSet.values.trustProxyHeader,
-    });
-    await initI18n();
-    await initApp(app);
-  } catch (error: unknown) {
-    console.log('Error while initializing app', error);
-    await envSet.poolSafe?.end();
```
Comment (xiaoyijun) [inline packages/core/src/index.ts:21]: It seems that we miss this logic in the new code.

File: packages/core/src/libraries/hook.ts (hunk 5 above)
Comment (gao-sun) [reply packages/core/src/libraries/hook.ts:37]: you are right, actually we can

Comment (gao-sun) [reply packages/core/src/libraries/hook.ts:37]: oh it doesn't have the built-in support yet, add LOG-4928

File: packages/core/src/index.ts (hunk 6 above)
Comment (gao-sun) [reply packages/core/src/index.ts:21]: you are right

**Pull request merged**